<!DOCTYPE html>
<html lang="en"><head>
<script src="09_nonparametrics_files/libs/clipboard/clipboard.min.js"></script>
<script src="09_nonparametrics_files/libs/quarto-html/tabby.min.js"></script>
<script src="09_nonparametrics_files/libs/quarto-html/popper.min.js"></script>
<script src="09_nonparametrics_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="09_nonparametrics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="09_nonparametrics_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="09_nonparametrics_files/libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.40">

  <meta name="author" content="Davide Risso">
  <title>Nonparametric and Bayesian methods</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="09_nonparametrics_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="09_nonparametrics_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="09_nonparametrics_files/libs/revealjs/dist/theme/quarto-bbe7401fe57d4b791b917637bb662036.css">
  <link href="09_nonparametrics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="09_nonparametrics_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="09_nonparametrics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="09_nonparametrics_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Nonparametric and Bayesian methods</h1>
  <p class="subtitle">Advanced Statistics and Data Analysis</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Davide Risso 
</div>
</div>
</div>

</section>
<section>
<section id="the-bootstrap" class="title-slide slide level1 center">
<h1>The Bootstrap</h1>

</section>
<section id="parametric-vs-nonparametric-inference" class="slide level2">
<h2>Parametric vs nonparametric inference</h2>
<p>Up until now, we have discussed <em>parametric models</em>, meaning that we have always made some assumptions on the data generating distribution.</p>
<div class="fragment">
<p>In particular, by assuming that the distribution belongs to a particular family, indexed by a parameter, we turn the difficult problem of estimating the distribution to the simpler problem of estimating a parameter.</p>
</div>
</section>
<section id="parametric-vs-nonparametric-inference-1" class="slide level2">
<h2>Parametric vs nonparametric inference</h2>
<p>Sometimes, we may not have enough information to assume a specific parametric form for our data generating distribution.</p>
<div class="fragment">
<p>Alternatively, we may want to obtain results that do not depend on the specific distribution that we assume.</p>
</div>
<div class="fragment">
<p>In such cases, we can rely on nonparametric techniques, such as <em>the bootstrap</em> and <em>permutations</em>.</p>
</div>
</section>
<section id="the-bootstrap-1" class="slide level2">
<h2>The Bootstrap</h2>

<img data-src="img/bootstrap_large.jpg" class="r-stretch quarto-figure-center"><p class="caption">The Surprising Adventures of Baron Munchausen (1781)</p></section>
<section id="frequentist-estimates" class="slide level2">
<h2>Frequentist estimates</h2>
<p>When we think about an estimator in statistics, we think about a <em>function of the data</em> that has some desirable properties in a <em>frequentist sense</em>.</p>
<div class="fragment">
<p>For instance, the distribution of an estimator can be interpreted in this way: if we repeat the same experiment an infinite amount of time we will get different estimate values, that taken together give us an idea of how variable my estimate is.</p>
</div>
</section>
<section id="frequentist-estimates-1" class="slide level2">
<h2>Frequentist estimates</h2>
<p>If we know the data generating distribution <span class="math inline">\(F(x)\)</span>, we can compute the standard deviation of the estimator by repeatedly sample new data <span class="math inline">\(x\)</span>.</p>
<div class="fragment">
<p><strong>In practice this is impossible because we do not know the distribution <span class="math inline">\(F\)</span>.</strong></p>
</div>
</section>
<section id="the-nonparametric-bootstrap" class="slide level2">
<h2>The (nonparametric) Bootstrap</h2>
<p>The main idea behind the nonparametric bootstrap is to substitute an estimate <span class="math inline">\(\hat{F}_n(x)\)</span> for the unknown <span class="math inline">\(F(x)\)</span> and then estimate the standard deviation by sampling multiple times from <span class="math inline">\(\hat{F}_n(x)\)</span>.</p>
<div class="fragment">
<p>This is done by creating a <em>bootstrap sample</em> <span class="math display">\[
x^* = (x^*_1, x^*_2, \ldots, x^*_n),
\]</span> from the original data <span class="math inline">\(x = (x_1, x_2, \ldots, x_n)\)</span>, where each <span class="math inline">\(x^*_i\)</span> is drawn randomly with equal probability and <em>with replacement</em> from <span class="math inline">\(\{x_1, x_2, \ldots, x_n\}\)</span>.</p>
</div>
</section>
<section id="the-nonparametric-bootstrap-1" class="slide level2 smaller scrollable">
<h2>The (nonparametric) Bootstrap</h2>
<p>Given a sample of <span class="math inline">\(n\)</span> observations, the bootstrap proceed as follows:</p>
<ol type="1">
<li class="fragment"><p>We independently generate a large number <span class="math inline">\(B\)</span> of bootstrap samples;</p></li>
<li class="fragment"><p>We compute for each bootstrap sample the estimate of the parameter, <span class="math inline">\(\hat \theta^*_b\)</span>, for <span class="math inline">\(b=1, \ldots, B\)</span>.</p></li>
<li class="fragment"><p>We compute the empirical standard deviation of the B <span class="math inline">\(\hat \theta^*_b\)</span> values:</p></li>
</ol>
<div class="fragment">
<p><span class="math display">\[
\hat{s}_{boot} = \sqrt{\frac{\sum_{b=1}^B (\hat \theta^*_b - \hat \theta^*)}{B-1}}, \quad \quad \text{ with } \quad \hat \theta^* = \sum_{b=1}^B \theta^*_b / B
\]</span></p>
</div>
</section>
<section id="motivation" class="slide level2 scrollable">
<h2>Motivation</h2>
<p>We notice that to obtain our estimate of <span class="math inline">\(\theta\)</span>, we proceed in two steps:</p>
<ol type="1">
<li class="fragment"><p><span class="math inline">\(x\)</span> is generated by sampling independently from the data generating distribution <span class="math inline">\(F(x);\)</span></p></li>
<li class="fragment"><p><span class="math inline">\(\hat \theta\)</span> is calculated from <span class="math inline">\(x\)</span> according to a certain function <span class="math inline">\(t\)</span>.</p></li>
</ol>
<div class="fragment">
<p>Schematically: <span class="math display">\[
F \to x \to \hat\theta.
\]</span></p>
</div>
</section>
<section id="motivation-1" class="slide level2 scrollable">
<h2>Motivation</h2>
<p>We don’t know <span class="math inline">\(F\)</span>, but we can estimate it by the <em>empirical distribution</em> of the observed data, <span class="math inline">\(\hat{F}_n\)</span>, which puts <span class="math inline">\(1/n\)</span> probability to each observed point <span class="math inline">\(x_i\)</span>:</p>
<ol type="1">
<li class="fragment"><p><span class="math inline">\(x^*\)</span> is generated by sampling independently from the empirical distribution <span class="math inline">\(\hat{F}_n(x);\)</span></p></li>
<li class="fragment"><p><span class="math inline">\(\hat \theta^*\)</span> is calculated from <span class="math inline">\(x^*\)</span> according to the same function <span class="math inline">\(t\)</span>.</p></li>
</ol>
<div class="fragment">
<p>Schematically: <span class="math display">\[
\hat{F}_n \to x^* \to \hat\theta^*.
\]</span></p>
</div>
</section>
<section id="motivation-2" class="slide level2">
<h2>Motivation</h2>
<p>In the real world we get to see <em>only one draw</em> <span class="math inline">\(x\)</span> from <span class="math inline">\(F\)</span> and hence only one value <span class="math inline">\(\hat\theta\)</span>.</p>
<div class="fragment">
<p>In the bootstrap world we can generate <em>as many draws as we want</em> from <span class="math inline">\(\hat{F}_n\)</span> and observe the distribution of <span class="math inline">\(\hat\theta^*.\)</span></p>
</div>
</section>
<section id="example-haplotype-frequencies" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<ul>
<li class="fragment"><p>A <em>haplotype</em> is a collection of alleles (DNA sequence variants) that are spatially adjacent on a chromosome, are usually inherited together (recombination tends not to disconnect them), and thus are genetically linked. In this case we are looking at linked variants on the Y chromosome.</p></li>
<li class="fragment"><p>We are interested in the frequencies of particular Y-haplotypes that consist of a set of different short tandem repeats (STR). The combination of STR numbers at the specific locations used for DNA forensics are labeled by the number of repeats at the specific positions.</p></li>
</ul>
</section>
<section id="example-haplotype-frequencies-1" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<p>The data look like this:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>  Individual DYS19 DXYS156Y DYS389m DYS389n DYS389p
1         H1    14       12       4      12       3
2         H3    15       13       4      13       3
3         H4    15       11       5      11       3
4         H5    17       13       4      11       3
5         H7    13       12       5      12       3
6         H8    16       11       5      12       3</code></pre>
</div>
</div>
<p>For instance, haplotype H1 has 14 repeats at position DYS19, 12 repeats at position DXYS156Y, etc.</p>
<p>Suppose we want to find the underlying proportion <span class="math inline">\(\theta\)</span> of a particular haplotype in a population of interest, by haplotyping <span class="math inline">\(n=300\)</span> men; and suppose we found H1 in <span class="math inline">\(y = 40\)</span> of them.</p>
</section>
<section id="parametric-approach" class="slide level2">
<h2>Parametric approach</h2>
<p>We can assume that the data generating distribution is <span class="math inline">\(Y \sim \mathcal{Bin}(n, \theta)\)</span>.</p>
<div class="fragment">
<p>By now we know that the <em>maximum likelihood estimate</em> of <span class="math inline">\(\theta\)</span> is <span class="math inline">\(\hat{\theta} = 40/300 = 0.13\)</span>.</p>
</div>
<div class="fragment">
<p>We can compute its standard error <span class="math display">\[
se(\hat\theta) = \sqrt{\hat\theta(1-\hat\theta)/n} = 0.0196
\]</span></p>
</div>
</section>
<section id="nonparametric-approach" class="slide level2">
<h2>Nonparametric approach</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href=""></a><span class="fu">set.seed</span>(<span class="dv">2422</span>)</span>
<span id="cb2-2"><a href=""></a></span>
<span id="cb2-3"><a href=""></a><span class="co"># observed data</span></span>
<span id="cb2-4"><a href=""></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">40</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">260</span>))</span>
<span id="cb2-5"><a href=""></a></span>
<span id="cb2-6"><a href=""></a>B <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb2-7"><a href=""></a>ts <span class="ot">&lt;-</span> <span class="fu">replicate</span>(B, {</span>
<span id="cb2-8"><a href=""></a>    xs <span class="ot">&lt;-</span> <span class="fu">sample</span>(x, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-9"><a href=""></a>    <span class="fu">mean</span>(xs)</span>
<span id="cb2-10"><a href=""></a>})</span>
<span id="cb2-11"><a href=""></a></span>
<span id="cb2-12"><a href=""></a><span class="fu">mean</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1343267</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href=""></a><span class="fu">sd</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.01936978</code></pre>
</div>
</div>
</section>
<section id="confidence-intervals" class="slide level2">
<h2>Confidence intervals</h2>
<p>The bootstrap distribution gives us a natural way to estimate the confidence interval by looking at the empirical quantiles of the bootstrap distribution</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href=""></a><span class="fu">quantile</span>(ts, <span class="fu">c</span>(<span class="fl">0.025</span>, <span class="fl">0.975</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      2.5%      97.5% 
0.09666667 0.17333333 </code></pre>
</div>
</div>
<div class="fragment">
<p>For comparison, the parametric approach gives</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href=""></a><span class="fu">binom.test</span>(<span class="dv">40</span>, <span class="dv">300</span>)<span class="sc">$</span>conf.int</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.09699221 0.17710876
attr(,"conf.level")
[1] 0.95</code></pre>
</div>
</div>
</div>
</section></section>
<section>
<section id="bayesian-inference" class="title-slide slide level1 center">
<h1>Bayesian inference</h1>

</section>
<section id="the-frequentist-phylosophy" class="slide level2 scrollable">
<h2>The Frequentist Phylosophy</h2>
<p>The statistical methods discussed so far are known as <em>frequentist</em>.</p>
<p>The frequentist point of view is based on the following ideas:</p>
<ol type="1">
<li class="fragment"><p><strong>Probabilities are objective properties of the real world</strong>: they are defined as the limit of the relative frequencies in infinite many trials.</p></li>
<li class="fragment"><p><strong>Parameters are fixed, unknown constants</strong>.</p></li>
</ol>
<div class="fragment">
<p>As a consequence, a frequentist methods are designed to have good <em>long-run frequency properties</em>.</p>
</div>
</section>
<section id="example" class="slide level2">
<h2>Example</h2>
<ul>
<li class="fragment"><p>A <span class="math inline">\(95\%\)</span> confidence interval is defined so that <em>it includes the true value of the parameter <span class="math inline">\(95\%\)</span> of the times if we repeat the experiment an infinite number of times</em>.</p></li>
<li class="fragment"><p><strong>Note that it is not the parameter but the interval to be random!</strong></p></li>
</ul>
</section>
<section id="the-bayesian-phylosophy" class="slide level2 scrollable">
<h2>The Bayesian Phylosophy</h2>
<p>The Bayesian approach takes a very different point of view:</p>
<ol type="1">
<li class="fragment"><p><strong>Probabilities describe a degree of belief</strong>: they are not objective properties of the world, but rather <em>subjective</em> propositions.</p></li>
<li class="fragment"><p><strong>Parameters are random variables</strong>: there is no such thing as the true parameter, all we know is the probability distribution of its values.</p></li>
</ol>
<div class="fragment">
<p>As a consequence, inference in a Bayesian setting takes the form of an “updated” distribution of the parameter: we can then extract point estimates and intervals from this distribution.</p>
</div>
</section>
<section id="model-specification" class="slide level2">
<h2>Model specification</h2>
<ul>
<li class="fragment"><p>In frequentist inference, we need a model for the data generating distribution (i.e., the likelihood).</p></li>
<li class="fragment"><p>In the Bayesian framework, in addition to the likelihood, we need to specify a <strong>prior distribution</strong> for the parameter(s).</p></li>
<li class="fragment"><p>This prior distribution <em>elicit</em> our prior belief in what the value of the parameter should be.</p></li>
<li class="fragment"><p>Inference is carried out by updating the prior distribution with the new knowledge that we observed in the data, to form the <strong>posterior distribution</strong>.</p></li>
</ul>
</section>
<section id="bayes-theorem" class="slide level2">
<h2>Bayes’ Theorem</h2>
<p>Recall Bayes’ Theorem as a way to “invert” the conditional probability: <span class="math display">\[
P(A | B) = \frac{P(B | A) P (A)}{P(B)}
\]</span></p>
</section>
<section id="example-births-in-paris" class="slide level2">
<h2>Example: births in Paris</h2>
<ul>
<li class="fragment"><p>Laplace was the first to formulate a statistical problem in Bayesian terms.</p></li>
<li class="fragment"><p>The question he poses was whether the probability of a female birth is or is not lower than 0.5.</p></li>
<li class="fragment"><p>Data: in Paris, from 1745 to 1770 there were 493,472 births, of which 241,945 were girls.</p></li>
</ul>
</section>
<section id="a-frequentist-solution" class="slide level2">
<h2>A frequentist solution</h2>
<ul>
<li class="fragment"><p>We can assume a binomial model and get a frequentist estimate for the probability of a female birth as <span class="math inline">\(\hat\theta = \frac{241,945}{493,472} = 0.4903\)</span></p></li>
<li class="fragment"><p>We can also compute a confidence interval <span class="math inline">\([0.4889, 0.4917]\)</span></p></li>
<li class="fragment"><p>Finally, we can test the null hypothesis <span class="math inline">\(H_0: \theta = 0.5\)</span> and reject it as the obtained p-value is close to 0.</p></li>
</ul>
</section>
<section id="a-bayesian-solution" class="slide level2">
<h2>A Bayesian solution</h2>
<ul>
<li class="fragment"><p>Laplace argues that he has no reason to <em>a priori</em> believe that any value of <span class="math inline">\(\theta\)</span> is more likely than others.</p></li>
<li class="fragment"><p>Hence, the <em>prior probability</em> is <span class="math inline">\(p(\theta) \sim \mathcal{U}(0, 1)\)</span>.</p></li>
<li class="fragment"><p>We can prove that in such case the <em>posterior distribution</em> is</p></li>
</ul>
<div class="fragment">
<p><span class="math display">\[
p(\theta | \text{data}) \propto \theta^{241,945} \, (1-\theta)^{493,472 - 241,945}
\]</span></p>
</div>
</section>
<section id="the-posterior-distribution" class="slide level2">
<h2>The posterior distribution</h2>

<img data-src="09_nonparametrics_files/figure-revealjs/unnamed-chunk-5-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:75.0%"></section>
<section id="bayes-theorem-revisited" class="slide level2 smaller">
<h2>Bayes’ Theorem revisited</h2>
<p>In statistics, we can rethink the Bayes’ Theorem in terms of the data and the parameter: <span class="math display">\[
p(\theta | y) = \frac{p(\theta) \, p(y | \theta)}{p(y)},
\]</span> where:</p>
<ul>
<li class="fragment"><span class="math inline">\(p(\theta)\)</span> is the <em>prior distribution</em> of the parameter</li>
<li class="fragment"><span class="math inline">\(p(y | \theta)\)</span> is the distribution of the data given a particular value of the parameter (i.e., <em>the likelihood</em>)</li>
<li class="fragment"><span class="math inline">\(p(y)\)</span> is the <em>marginal distribution of the data</em> (across all possible values of the parameter)</li>
</ul>
<div class="fragment">
<div class="callout callout-note callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Note</strong></p>
</div>
<div class="callout-content">
<p>Much of Bayesian statistics deals with ways to compute/approximate <span class="math inline">\(p(y)\)</span>.</p>
</div>
</div>
</div>
</div>
</section>
<section id="example-haplotype-frequencies-2" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<ul>
<li class="fragment"><p>A <em>haplotype</em> is a collection of alleles (DNA sequence variants) that are spatially adjacent on a chromosome, are usually inherited together (recombination tends not to disconnect them), and thus are genetically linked. In this case we are looking at linked variants on the Y chromosome.</p></li>
<li class="fragment"><p>We are interested in the frequencies of particular Y-haplotypes that consist of a set of different short tandem repeats (STR). The combination of STR numbers at the specific locations used for DNA forensics are labeled by the number of repeats at the specific positions.</p></li>
</ul>
</section>
<section id="example-haplotype-frequencies-3" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<p>The data look like this:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>  Individual DYS19 DXYS156Y DYS389m DYS389n DYS389p
1         H1    14       12       4      12       3
2         H3    15       13       4      13       3
3         H4    15       11       5      11       3
4         H5    17       13       4      11       3
5         H7    13       12       5      12       3
6         H8    16       11       5      12       3</code></pre>
</div>
</div>
<p>For instance, haplotype H1 has 14 repeats at position DYS19, 12 repeats at position DXYS156Y, etc.</p>
<p>Suppose we want to find the underlying proportion <span class="math inline">\(\theta\)</span> of a particular haplotype in a population of interest, by haplotyping <span class="math inline">\(n=300\)</span> men; and suppose we found H1 in <span class="math inline">\(y = 40\)</span> of them.</p>
</section>
<section id="frequentist-approach" class="slide level2">
<h2>Frequentist approach</h2>
<p>We can assume that the data generating distribution is <span class="math inline">\(Y \sim \mathcal{Bin}(n, \theta)\)</span>, with <span class="math inline">\(\theta\)</span> fixed and unknown.</p>
<div class="fragment">
<p>By now we know that the <em>maximum likelihood estimate</em> of <span class="math inline">\(\theta\)</span> is <span class="math inline">\(\hat{\theta} = 40/300 = 0.13\)</span>.</p>
</div>
<div class="fragment">
<p>We can compute confidence intervals and test hypotheses on the true value <span class="math inline">\(\theta\)</span>.</p>
</div>
</section>
<section id="bayesian-approach" class="slide level2">
<h2>Bayesian approach</h2>
<ul>
<li class="fragment"><p>In a Bayesian setting, <span class="math inline">\(\theta\)</span> is not fixed, and we need to assume a <em>prior</em> probability distribution for the parameter as well.</p></li>
<li class="fragment"><p>In principle we could use <em>any</em> distribution.</p></li>
<li class="fragment"><p>It is convenient to specify a <em>Beta distribution</em>, i.e.</p></li>
</ul>
<div class="fragment">
<p><span class="math display">\[
p_{\alpha, \beta}(x) = \frac{x^{\alpha-1} (1-x)^{\beta-1}}{B(\alpha, \beta)},
\]</span> where <span class="math inline">\(B(\alpha, \beta) = \frac{\Gamma(\alpha) \Gamma(\beta)}{\Gamma(\alpha+\beta)}\)</span> is called the <em>beta function</em>.</p>
</div>
</section>
<section id="prior-distribution" class="slide level2">
<h2>Prior distribution</h2>
<ul>
<li class="fragment"><p>We can see how the prior distribution depends on two parameters, in this case <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span>.</p></li>
<li class="fragment"><p>These are known as <em>hyper-parameters</em> and are typically chosen according to some prior information.</p></li>
<li class="fragment"><p>The choice of a Beta distribution in this case is mathematically convenient, as one can show that the posterior distribution will be again a Beta distribution.</p></li>
<li class="fragment"><p>We say that the Beta distribution is the <em>conjugate prior</em> for the binomial model.</p></li>
</ul>
</section>
<section id="how-to-choose-the-hyperparameters" class="slide level2">
<h2>How to choose the hyperparameters</h2>
<ul>
<li class="fragment"><p>The hyperparameters are our way to express our prior knowledge of the problem.</p></li>
<li class="fragment"><p>An intuitive way is to think about the prior mean and variance as a way to elicit what is the value that we expect before performing the data (mean) and how confident we are in our prediction (variance).</p></li>
<li class="fragment"><p>The Beta distribution is quite flexible in that it includes the <em>uniform distribution</em> (absence of prior information) as well as very concentrated distributions.</p></li>
</ul>
</section>
<section id="how-to-choose-the-hyperparameters-1" class="slide level2">
<h2>How to choose the hyperparameters</h2>
<p>If <span class="math inline">\(X \sim \mathcal{Beta}(\alpha, \beta)\)</span>, then <span class="math display">\[
E[X] = \frac{\alpha}{\alpha + \beta}
\]</span> and</p>
<p><span class="math display">\[
Var(X) = \frac{\alpha\beta}{(\alpha + \beta)^2(\alpha+\beta+1)}
\]</span></p>
</section>
<section id="example-haplotype-frequencies-4" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<p>Let’s think about three situations:</p>
<ul>
<li class="fragment"><p>We have no prior knowledge on the haplotype frequencies in the population: we can choose a uniform distribution.</p></li>
<li class="fragment"><p>We know from the literature that one in five men should have haplotype H1, but we are not very confident.</p></li>
<li class="fragment"><p>We are very confident that one in five men has haplotype H1.</p></li>
</ul>
</section>
<section id="example-prior-distribution" class="slide level2">
<h2>Example: Prior distribution</h2>

<img data-src="09_nonparametrics_files/figure-revealjs/unnamed-chunk-7-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:75.0%"></section>
<section id="the-posterior-distribution-1" class="slide level2">
<h2>The posterior distribution</h2>
<p>If the prior is Beta and the likelihood is Binomial, the posterior distribution is still Beta, but with updated parameters: <span class="math display">\[
p(\theta | y) \sim \mathcal{Beta}(\alpha + y, \beta + n - y).
\]</span></p>
<div class="fragment">
<p>Hence, <span class="math display">\[
E[\theta | y] = \frac{\alpha+y}{\alpha + \beta + n},
\]</span></p>
<p>which is a synthesis of the prior information and the information contained in the data.</p>
</div>
</section>
<section id="the-posterior-distribution-2" class="slide level2">
<h2>The posterior distribution</h2>

<img data-src="09_nonparametrics_files/figure-revealjs/unnamed-chunk-8-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:75.0%"></section>
<section id="the-posterior-distribution-3" class="slide level2">
<h2>The posterior distribution</h2>

<img data-src="09_nonparametrics_files/figure-revealjs/unnamed-chunk-9-1.png" class="quarto-figure quarto-figure-center r-stretch" style="width:75.0%"><ul>
<li class="fragment"><p>Note that we do not have a single answer for the “best” value of <span class="math inline">\(\theta\)</span> but a distribution.</p></li>
<li class="fragment"><p>However, we can calculate the <em>posterior mean</em> and <em>probabilities</em> based on the posterior distribution, e.g., <span class="math inline">\(Pr(\theta &gt; 0.2) \approx 0.003\)</span>.</p></li>
</ul>
</section>
<section id="the-posterior-distribution-4" class="slide level2">
<h2>The posterior distribution</h2>
<p>Let’s look again at the posterior distribution: <span class="math display">\[
\theta | y \sim \mathcal{Beta}(\alpha + y, \beta + n - y).
\]</span></p>
<ul>
<li class="fragment"><p>As we increase the sample size, we give more weight to the data.</p></li>
<li class="fragment"><p>Conversely, if we increase the hyperparameters we give more weight to our prior belief.</p></li>
</ul>
</section>
<section id="credibility-intervals" class="slide level2">
<h2>Credibility intervals</h2>
<ul>
<li class="fragment"><p>The Bayesian analog of the <em>confidence intervals</em> are the <strong>credibility intervals</strong>.</p></li>
<li class="fragment"><p>These are much more natural to think about: they are straightforward to interpret in terms of the posterior probability of the parameter.</p></li>
<li class="fragment"><p>For instance, we can use quantiles to compute the range in which the parameter lies with <span class="math inline">\(95\%\)</span> probability.</p></li>
</ul>
<div class="fragment">
<p><span class="math display">\[
Pr(q_{2.5} \leq \theta \leq q_{97.5}) = 0.95
\]</span></p>
</div>
</section>
<section id="example-haplotype-frequencies-5" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<p>Putting it all together, we can specify our Bayesian model as:</p>
<ul>
<li class="fragment"><p>The data generating distribution is <span class="math inline">\(y|\theta \sim \mathcal{Bin}(n=300, \theta);\)</span></p></li>
<li class="fragment"><p>The prior distribution is <span class="math inline">\(\theta \sim \mathcal{Beta}(\alpha = 16, \beta = 64);\)</span></p></li>
</ul>
<div class="fragment">
<p>Observing <span class="math inline">\(y=40\)</span> leads to the posterior distribution <span class="math display">\[
\theta | y \sim \mathcal{Beta}(16+40, 64 + 300 - 40).
\]</span></p>
</div>
</section>
<section id="example-haplotype-frequencies-6" class="slide level2">
<h2>Example: Haplotype frequencies</h2>
<p>The posterior mean is</p>
<p><span class="math display">\[
E[\theta | y] = \frac{\alpha+y}{\alpha + \beta + n} = 0.147
\]</span></p>
<p>(compared to a MLE of <span class="math inline">\(\hat\theta = 0.13\)</span>).</p>
<div class="fragment">
<p>In addition, the 95% credibility interval is</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href=""></a><span class="fu">qbeta</span>(<span class="fu">c</span>(<span class="fl">0.025</span>, <span class="fl">0.975</span>), <span class="at">shape1 =</span> <span class="dv">16</span><span class="sc">+</span><span class="dv">40</span>, <span class="at">shape2 =</span> <span class="dv">64</span><span class="sc">+</span><span class="dv">300-40</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1135840 0.1846664</code></pre>
</div>
</div>
</div>
</section>
<section id="not-just-beta" class="slide level2">
<h2>Not just Beta</h2>
<ul>
<li class="fragment"><p>Many of the concepts that we have seen in this example extend to many other distributions!</p></li>
<li class="fragment"><p>For instance if we have Gaussian data, a Gaussian prior on the mean will lead to a Gaussian posterior.</p></li>
<li class="fragment"><p>This is true for the linear model too!</p></li>
</ul>
</section>
<section id="not-just-conjugate-priors" class="slide level2">
<h2>Not just conjugate priors</h2>
<p>There are many different choices of prior distributions, and when closed form solutions do not exist to compute the posterior, we need to use numerical approximations or simulation approaches.</p>
<p>Markov Chain Monte Carlo (MCMC) and Variational Approximations are perhaps the two most popular techniques.</p>

</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="09_nonparametrics_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="09_nonparametrics_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="09_nonparametrics_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="09_nonparametrics_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>